EmptySet with S

dnabanita7 opened the pull request:
Since,Singleton classes have instances that are the same name as the class.So,
EmptySet should return its instance rather than its class name.
As well,I changed the test_args.py to test if S.EmptySet is EmptySet().
Fixes #16337 

<!-- BEGIN RELEASE NOTES -->
* sets
   * changed __init__.py to show instance of EmptySet when being called rather than showing its classname
<!-- END RELEASE NOTES -->

dnabanita7 committed:
Changed subs() in matrices for iterable objects

The zip object is iterable.subs method in common.py didn't have a case for iterable objects.
Still I have to change the subs method in core.basic. The isinstance doesn't track iterable objects which don't have _getitem_ module,zip not having one creates an error. So,Shall I change isinstance() check to iter() in subs module in core.basic.

Almost Fixes #10589

I have added tests for the same.

dnabanita7 committed:
Update test_commonmatrix.py

dnabanita7 committed:
Update common.py

dnabanita7 committed:
Changed importing of EmptySet

sympified EmptySet

Naba7 commented:
I need help resolving issues in union.py as well as in intersection.py. Even if I change,more errors is being shown

oscargus commented:
I think that you may want to import the type, not the object, here.

Naba7 commented:
type as in I will add a check for type of object?

Naba7 commented:
sorry I couldn't get you. here means in union.py and intersection.py? @oscargus

oscargus commented:
If I understand it correctly (not sure that I do), the line: `@dispatch(EmptySet, Set)` (which causes the error) means that that method should be applied if the arguments are of the **type** `EmptySet` and `Set`, respectively. Now, with your change `EmptySet` is an **object** (`S.EmptySet`) rather than a type, the `@dispatch`-decorator will error. I think that it may be solved by figuring out how to refer to the class `EmptySet`. Probably, `EmptySet.__class__`will do.

Also, please remove the added empty lines in the other files.

oscargus commented:
I think it is enough to import `EmptySet` from `sympy.sets.sets` in the problematic files (and not from `sympy`) and it should be OK (compare with how `UniversalSet` is handled in union.py and intersection.py).

dnabanita7 committed:
Updated object type to class type

changed EmptySet() to EmptySet.__class__() to get the class type rather than object
type after changed in __init__.py of sets.sets

Naba7 commented:
@asmeurer @oscargus Can you give a look ? I think it is ready to go

dnabanita7 committed:
updated some changes

changes

dnabanita7 committed:
changes to switch types

changed object type of EmptySet to class type

Naba7 commented:
I have created a pr in symengine.py changing the emptyset object to class type in https://github.com/symengine/symengine.py/pull/275

isuruf commented on sympy/utilities/tests/test_wester.py:
```diff
@@ -971,7 +971,7 @@ def test_M19():
 
 def test_M20():
-    assert solveset(sqrt(x**2 + 1) - x + 2, x) == EmptySet()
+    assert solveset(sqrt(x**2 + 1) - x + 2, x) == EmptySet.__class__()
```
Why are you checking for the class here?

Naba7 replied:
Well the function that is calling needs the instance of the class . So here class of EmptySet is required rather than EmptySet

smichr replied:
Can you just write `S.EmptySet`? That what we do with something like testing that x is S(1):  `assert x is S.One`.

Naba7 replied:
Well I have changed it to`` EmptySet`` because now it is the instance

oscargus commented:
It is not clear if it actually is required to change in all these places. Basically you can choose to import EmptySet from sympy to get the object or sympy.sets.sets to get the class

In addition, S.EmptySet == S.EmptySet.__class__(), so there should basically not be any calls to __class__() as this only brings the object back again. __class__ should only be used to get the class (if, for some reason, EmptySet is not imported from sympy.sets.sets)

Naba7 commented:
Well I have tried it.

asmeurer commented on sympy/solvers/solveset.py:
```diff
@@ -3051,7 +3051,7 @@ def nonlinsolve(system, *symbols):
         # positive dimensional system
         res = _handle_positive_dimensional(polys, symbols, denominators)
-        if isinstance(res, EmptySet) and any(not p.domain.is_Exact for p in polys):
+        if isinstance(res, EmptySet.__class__) and any(not p.domain.is_Exact for p in polys):
```
Use `type(EmptySet)` to get the type.

Naba7 replied:
alright

asmeurer commented:
`EmptySet.__class__()` should just be `EmptySet`, since that is now the instance.

asmeurer commented:
IMHO we should move the lines that are in `__init__.py` to `sets.py` so that even the specialized import gives the instance (otherwise it is confusing). You can always get the class with `type`.

asmeurer commented:
The release notes should specify that `EmptySet` is now the instances instead of the class. Implementation details such as `__init__.py` do not need to go in the release notes entry.

Naba7 commented:
> `EmptySet.__class__()` should just be `EmptySet`, since that is now the instance.

Alright I shall revert all changes to ``EmptySet`` rather than ``EmptySet.__class__()``

Naba7 commented:
> IMHO we should move the lines that are in `__init__.py` to `sets.py` so that even the specialized import gives the instance (otherwise it is confusing). You can always get the class with `type`.

Yes,we should do this otherwise importing from .sets and sympt is creating errors and is confusing.

Naba7 commented:
if we use ``EmptySet`` rather than ``EmptySet.__class__()`` then I think symengine doesn't need any change in their files.

Naba7 commented:
@asmeurer  actually the thing in here is all the files go through a decorator ``@dispatcher`` and it directs the pointer to ``multipledispatch/dispatcher.py`` which calls the classes of functions . Hence it is necessary to call class of EmptySet rather than instance of it. I will change ``EmptySet.__class__`` to ``type(EmptySet)``

Naba7 commented:
And what changes should I do with  ``__init__.py()``?

dnabanita7 committed:
changed instance of EmptySet

to get the class of EmptySet we use type(EmptySet)

isuruf commented:
You also need to change the printing of EmptySet

Naba7 commented:
> You also need to change the printing of EmptySet

Can you be more descriptive please?

isuruf commented:
EmptySet is printed as `EmptySet()`, but this has to be changed to `EmptySet`

Naba7 commented:
That is the problem with UniversalSet too as said by @asmeurer . Because it is already stored as ``EmptySet()`` in ``_instances``

Naba7 commented:
@isuruf I am not finding a way to change ``EmptySet()`` to ``EmptySet``

asmeurer commented:
The printer should be in sympy/printing/str.py

Naba7 commented:
Actually I want to know where ``set()`` is defined .because the classes in ``set()`` are printed as same as in there. So, changing the contents of ``set()`` will change the printing of ``EmptySet`` and ``UniversalSet``. But that will create a problem as ``all_classes = set()`` , ``all_classes`` is set to ``set()`` and it will store the ``instances`` rather than `` classes`

Naba7 commented:
I am busy with my exams . Can I send a pr right after it is done?

oscarbenjamin commented:
> I am busy with my exams . Can I send a pr right after it is done?

Yes of course. Good luck with the exams.

Naba7 commented:
thanks

oscarbenjamin commented:
This has merge conflicts now. Are you still working on this?

Naba7 commented:
No,I couldn't resolve this
